[Woo POS] Coupons: Pagination - Wrap Up

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR aims to wrap up pagination functionality with coupons. 

- Updated page tracking and `hasMorePages` logic
- Added loading state for pagination
- Added error state for pagination
- Stopped reloading coupons and products each time the tab is opened, only when there's no data
- Started preloading coupons when POS is opened

### Has more pages check

`PointOfSaleCouponsService` uses `CouponStore.synchronizeCoupons`, which already has built-in pagination check functionality. It doesn't use `X-WP-TotalPages` and checks if the returned result is smaller or equal to the page size. If it's smaller, it means there are no pages to load. 

At the moment,  `PointOfSaleCouponsService` didn't use the result of `synchronizeCoupons` so there were infinite page loading issues.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Prerequisites:
- Enable `enableCouponsInPointOfSale` feature flag.
- Have a store with more than 25 coupons or decrease `defaultPageSize` in `PointOfSaleCouponService`

## Happy Path
1. Open POS
2. Switch to Coupons
3. Confirm the data was already preloaded
4. Scroll down
5. Confirm the loading state is shown at the bottom while the next page is loading
6. Confirm the data loads
7.  Switch between tabs
8. Confirm the data is not reloaded

## Unhappy Path
1. Open POS
2. Switch to Coupons
3. Turn off network / simulate another type of issue
4. Scroll down
5. Confirm the error state is shown under the last row
6. Turn on network
8. Tap retry
9. Confirm the next page loads
10. Switch between tabs
11. Confirm the data is not reloaded

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

Tested on iPad Air 11 Inch M3 iOS 18.4
Confirmed that there's no more infinite loading, and network requests are not made when there are no more pages to load.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/416a6761-c0f1-4c5c-a2ac-7cbadd6844ab

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Enable pagination tracking for Coupons by using synchronizeCoupons result

synchronizeCoupons returns `true` if the are more pages to sync. Use this result in `providePointOfSaleCoupons` to return a correct 'hasMorePages' value of `PagedItems`

staskus committed:
Resync Coupins paginationTracker when loading/reloading the first page

staskus committed:
Add loading state to pagination

staskus committed:
Don't reload data when switching between Products and Coupons

No need to reload data each time the tabs are switched. Reloading data breaks pagination. Pull to Refresh can be used for this.

staskus committed:
Limit fetching local coupons data to defaultPageSize of the remote request

Local data gets quickly replaced by remote results. Matching the page size for local and remote results avoids some issues, such as disappearing items if users scrolls quickly after the initial load.

staskus committed:
Handle coupon pagination errors

staskus committed:
Load coupons when POS is opened

staskus committed:
Fix typo in the comment within PointOfSaleCouponService

staskus committed:
Merge branch 'trunk' into woomob-305-woo-pos-coupons-pagination-infinite-page-loading-isssue

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -102,6 +102,9 @@ struct PointOfSaleDashboardView: View {
         }
         .task {
             await posModel.purchasableItemsController.loadItems(base: .root)
+            if ServiceLocator.featureFlagService.isFeatureFlagEnabled(.enableCouponsInPointOfSale) {
```
Well remembered

joshheald commented on Yosemite/Yosemite/PointOfSale/PointOfSaleCouponService.swift:
```diff
@@ -51,31 +51,37 @@ public final class PointOfSaleCouponService: PointOfSaleCouponServiceProtocol {
         do {
-            try await syncCouponsFromRemote(pageNumber: pageNumber)
+            let hasMorePages = try await syncCouponsFromRemote(pageNumber: pageNumber)
+            let coupons = try await provideLocalPointOfSaleCoupons(limit: nil)
```
Perhaps this could use some comments to explain exactly what we do? It's a little confusing to sync from the remote, then provide the local coupons if that succeeds.

In the context of this PR I can understand it but I can see people getting thrown by it in future.

joshheald commented on Yosemite/Yosemite/PointOfSale/PointOfSaleCouponService.swift:
```diff
@@ -95,23 +101,23 @@ public final class PointOfSaleCouponService: PointOfSaleCouponServiceProtocol {
-            return .init(items: posItems, hasMorePages: true)
+            return mapCouponsToPOSItems(coupons: storageCoupons)
         } catch {
             debugPrint("Failed to load coupons from storage:", error)
```
Not in the PR, but shouldn't this use DDLogError?

joshheald reviewed (approved):
Works well, :shipit: 

I've left a note mentioning an area that could use some more explanation, IMO. It's related to the order of fetching from the remote then showing local results, but also that there are other calls to show local results from outside the service.

The locally persisted data is only really used while waiting for a network request – once that times out or errors, the fullscreen error is shown instead. Perhaps a different error state would be more appropriate, like a banner at the top?

With the move to pre-loading coupons, we probably get less benefit from using persisted data at all, right now. Maybe when we use persisted products, we can get the improvement by removing the fullscreen loading indicator and just going straight in to POS. It's worth getting right for that, at least

We don't support offline mode, so I hesitate to mention it... but at the same time, it's good to be sure that the effort and complexity cost of persistence is actually giving a benefit.

staskus commented:
@joshheald thanks for the good comments.

> With the move to pre-loading coupons, we probably get less benefit from using persisted data at all, right now.

You're right. Pre-loading has removed the benefit of local-first fetching. Also, the remote `synchronize` call wipes out local storage data when we're fetching the first page. So, we only load the first page from the storage as well. I understand why it's done this way. If new items are added or older ones are updated, some data in the middle can get outdated or skipped. One of the solutions is a background synchronization mechanism that would update all the coupons in the storage, perhaps only fetching the newest ones by a modified date. Similar to what we discussed about the products. 

> We don't support offline mode, so I hesitate to mention it... but at the same time, it's good to be sure that the effort and complexity cost of persistence is actually giving a benefit.

I'll think if there's a reasonable mechanism we could apply now to get the benefits out of local. 

Having local did not cost much more effort since we reused what we had for IPP. But I agree that currently, we could have a remote-only solution that would produce almost identical UX.  

> The locally persisted data is only really used while waiting for a network request – once that times out or errors, the fullscreen error is shown instead. Perhaps a different error state would be more appropriate, like a banner at the top?

Good idea. One of the ways to get the UX benefit out of a local-first approach. I wonder if something similar to what we have with pagination error could be reused, just on the top.

staskus committed:
Replace debugPrint with DDLogError

staskus committed:
Add comments to PointOfSaleCouponService explaining local-first pagination

staskus commented:
> Perhaps a different error state would be more appropriate, like a banner at the top?

@joshheald , reusing `.inlineError` for the header would look like this:

![Simulator Screenshot - iPad Air 11-inch (M3) - 2025-04-16 at 11 27 28](https://github.com/user-attachments/assets/fd44d9ad-0047-4499-8579-c6052b7a7d03)

We could do the same if product pull to refresh fails. As with pagination, no value in removing all the products if a single action fails.

joshheald commented:
Yeah. For context, we discussed this when doing Variation support, and decided against it. 

I still think it's an improvement and like it, either adding a new state for `refreshError` or adding a position parameter to the existing one.

Android may find it trickier, I'm not sure, or it may just be that they didn't have time in the variations project.

staskus commented:
> Yeah. For context, we discussed this when doing Variation support, and decided against it.

@joshheald got it. It would be easy to do for products as well if we decide to do it. I could bring it up for a design input as well. 

> I still think it's an improvement and like it, either adding a new state for refreshError or adding a position parameter to the existing one.

Quite easy if we include context into `inlineError`. `.refresh` or `.pagination`.

staskus committed:
Add context to ItemListState inlineError

The request with items may fail when paginating and refreshing the list. A different context may require a different presentation.

In case of ItemList, presenting:
- .pagination errors in the footer
- .refresh errors in the header

staskus committed:
Present couponsRefreshError when coupons loading fails while items are loaded

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/242">22.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Merge branch 'trunk' into woomob-305-woo-pos-coupons-pagination-infinite-page-loading-isssue

staskus committed:
Linter fixes in PointOfSaleCouponsControllerTests
